### PR TITLE
create install script

### PR DESCRIPTION
**Pull request by DanielSGH**
- [x] Check if firefox and git are installed
- [x] Clone repo
- [x] Move chrome folder to profile folder root
- [ ] Detect profile folder automatically (deducted from files in mozilla/firefox)
- [ ] Options for recommended settings (may do in separate PR)

Designed to make the installation a lot simpler, by just using `curl` to execute the script

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,6 +5,18 @@ A Firefox userChrome.css theme that aims to recreate the look and feel of the Ch
 <img src="screenshots/thumbnail.png" alt="thumbnail screenshot" title="Screenshot taken with macOS Monterey / Firefox Nightly 96.0a1 (2021-11-30)" width="800">
 
 ## How to install
+### Via Linux and MacOS shell script
+Paste this into your terminal emulator:
+```sh
+# Linux may not have $TMPDIR set by default
+if [ -z $TMPDIR ]; then TMPDIR=/tmp; fi;
+
+curl https://raw.githubusercontent.com/bmFtZQ/edge-frfox/blob/main/install.sh > $TMPDIR/installer.sh && chmod +x $TMPDIR/installer.sh && sh $TMPDIR/installer.sh
```
Comment (DanielSGH) [inline README.md:14]: FYI this won't work until it exists in the main branch, for now you can test it with:

```sh
# Linux may not have $TMPDIR set by default
if [ -z $TMPDIR ]; then TMPDIR=/tmp; fi;

curl https://raw.githubusercontent.com/bmFtZQ/edge-frfox/bcf1ea15fd7cecddeda0e3cadde5c5626b9d8da0/install.sh > $TMPDIR/installer.sh && chmod +x $TMPDIR/installer.sh && sh $TMPDIR/installer.sh
```

**Comment by DanielSGH** [comment]
Might require testing on MacOS, will probably do a Windows .ps1 script later

**Comment by DanielSGH** [comment]
@bmFtZQ This is probably all I'll do for now, let me know what you think

File: install.sh [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    delete_pref $pref;
+  done;
+
+  echo "uninstall complete."
+  exit 0;
+fi
+
+git --version 2>&1 > /dev/null;
+if [ ! $? -eq 0 ]; then
+  echo "ERROR: git is not installed... Please install it."
```
Comment (bmFtZQ) [inline install.sh:84]: I think the theme might be more useful if it didn't use git as it is not installed by default on Windows or macOS (and maybe some Linux distributions?). 

URL for downloading the tarball of the current latest verison:
https://github.com/bmFtZQ/edge-frfox/archive/refs/heads/main.tar.gz

**Review by bmFtZQ — CHANGES_REQUESTED** [review]
Sorry for taking so long to respond this!

I already had an update script [here](https://github.com/bmFtZQ/userchrome-update-script/) but it requires the user to manually enter the profile path.

I think your installation script is better but it has some issues/changes that could be made to improve it:
* Don't modify `prefs.js`, instead use the `user.js` file which can be used to enable options automatically.
* The commands used to execute the script could be shortened to make installation more convenient (see the file comments for more details).
* Maybe use curl to download a tarball of the repository instead of using git as it isn't installed by default on most systems. The URL for downloading a tarball of the repo is: https://github.com/bmFtZQ/edge-frfox/archive/refs/heads/main.tar.gz

File: README.md [hunk 3]
```diff
@@ -5,6 +5,29 @@ A Firefox userChrome.css theme that aims to recreate the look and feel of the Ch
 <img src="screenshots/thumbnail.png" alt="thumbnail screenshot" title="Screenshot taken with macOS Monterey / Firefox Nightly 96.0a1 (2021-11-30)" width="800">
 
 ## How to install
+### Via Linux and MacOS shell script
+Paste this into your terminal emulator:
+```sh
+# Linux may not have $TMPDIR set by default
+TMP_DIR="${TMPDIR:-$(dirname $(mktemp))}"
+
+curl https://raw.githubusercontent.com/bmFtZQ/edge-frfox/create-install-script/install.sh > $TMP_DIR/installer.sh && chmod +x $TMP_DIR/installer.sh && sh $TMP_DIR/installer.sh
```
[the remark this answers is not in the document]
Comment (DanielSGH) [reply README.md:16]: adding `-s uninstall` to the end of that command seems to work perfectly for now

**Comment by DanielSGH** [comment]
I added a better way to uninstall, which preserves the user's `user.js` in case they want to keep it but removes only the custom settings added by the script if the user wishes.

I've also done everything you suggested above, let me know if there's anything you'd change.

I've also updated the README.md accordingly.

File: install.sh [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  echo "uninstall complete."
+  exit 0;
+fi
+
+#################
+# INSTALL PHASE #
+#################
+
+echo "Installing..."
+if ! curl -L "https://github.com/bmFtZQ/edge-frfox/archive/refs/heads/main.tar.gz" | tar xz -C $TMP_DIR --strip-components=1 --one-top-level=$PROJECT_NAME; then
```
Comment (bmFtZQ) [inline install.sh:125]: Causes error on macOS as `--one-top-level` is not supported in the macOS version of `tar`.

When using `mktemp -d` the files can be placed directly inside the temp directory.

File: install.sh [hunk 5]
```diff
@@ -0,0 +1,153 @@
+#!/bin/bash
+
+# VARIABLES, CHANGE AS NEEDED
+GITHUB_REPO="https://github.com/bmFtZQ/edge-frfox.git"
+PROJECT_NAME=$(basename $GITHUB_REPO | cut -d '.' -f 1)
+TMP_DIR="${TMPDIR:-$(dirname $(mktemp))}"
+
+if [[ $OSTYPE == "darwin"* ]]; then
+  FIREFOX_DIR=$HOME/Library/Application\ Support/Firefox
```
Comment (bmFtZQ) [inline install.sh:9]: Use double quotes here for the path, also would probably be best if double quotes are used for all commands involving the profile or Firefox path due to the space in the path name.

File: install.sh [hunk 6]
```diff
@@ -0,0 +1,153 @@
+#!/bin/bash
+
+# VARIABLES, CHANGE AS NEEDED
+GITHUB_REPO="https://github.com/bmFtZQ/edge-frfox.git"
+PROJECT_NAME=$(basename $GITHUB_REPO | cut -d '.' -f 1)
+TMP_DIR="${TMPDIR:-$(dirname $(mktemp))}"
```
Comment (bmFtZQ) [inline install.sh:6]: Is there any reason `mktemp -d` can't be used on its own here to make a new temp directory?

```sh
TMP_DIR="$(mktemp -d)"
```

This seems to work better on macOS.

File: install.sh [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+#################
+
+echo "Installing..."
+if ! curl -L "https://github.com/bmFtZQ/edge-frfox/archive/refs/heads/main.tar.gz" | tar xz -C $TMP_DIR --strip-components=1 --one-top-level=$PROJECT_NAME; then
+  echo -e "${RED}Installation failed: ${NC}Failed to get github repo tarball and extract it";
+  exit 0;
+fi
+
+echo "Copying theme folder...";
+cp -r $TMP_DIR/$PROJECT_NAME/chrome $PROFILE_ROOTDIR;
```
Comment (bmFtZQ) [inline install.sh:131]: When using `mktemp -d` and removing `--one-top-level` the `$PROJECT_NAME` variable can be removed here.

File: install.sh [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+# OPTIONAL SETTINGS #
+#####################
+
+echo "Optional settings, refer to https://github.com/bmFtZQ/edge-frfox/tree/main#how-to-install";
+for ((i = 0; i < ${#OPTIONALS[@]}; i += 2)); do
+  ans="n"
+  read -e -p "Set ${OPTIONALS[i]} to ${OPTIONALS[i+1]?} [y/N]: " in
+  ans="${in:-$ans}";
+
+  if [[ ! ${ans,,} =~ ^(yes|y)$ ]]; then
```
Comment (bmFtZQ) [inline install.sh:144]: The lowercase parameter expansion `${var,,}` is not supported in bash 3.2.

Maybe something like this can be used: https://stackoverflow.com/a/2264451

```sh
ans="$(echo "$ans" | tr "[:upper:]" "[:lower:]")"
if [[ ! ${ans} =~ ^(yes|y)$ ]]; then
  continue;
fi
```

This would also need to be applied in other areas of the file.

**Review by bmFtZQ — CHANGES_REQUESTED** [review]
Thanks for creating these scripts and making changes to them, I have a few issues to report and some possible solutions (See the file comments for more info).

Also with the `user.js` file, if running the script again after installing it, duplicate values will be added to the `user.js` file as Firefox will never edit `user.js`.

When Firefox is opened, it sets the preferences found in `user.js`. Removing them from the `user.js` file later does not remove the preferences from Firefox as they have already been set previously.

I think the `user.js` file should just be completely overwritten by copying it on install or deleted on uninstall. This would also resolve the encoding error with Power Shell (see file comments for more info).

File: install.ps1 [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+Expand-Archive -Path "$env:temp\$PROJECT_NAME.zip" -DestinationPath "$env:temp" -Force
+if (!($?)) {
+  Write-Host "ERROR: Failed to extract $env:temp\$PROJECT_NAME.zip." -ForegroundColor Red;
+  Write-Host "Message: $($Error[0].Exception.Message)";
+  exit 0;
+}
+
+echo "Copying theme folder..."
+Copy-Item -Recurse -Path "$env:temp\$PROJECT_NAME-main\chrome" -Destination $PROFILE_ROOTDIR
+Get-Content "$env:temp\$PROJECT_NAME-main\user.js" | Out-File "$PROFILE_ROOTDIR\user.js" -append -Encoding UTF8 | Out-Null
```
[the remark this answers is not in the document]
Comment (DanielSGH) [reply install.ps1]: I can't seem to reproduce this on my windows 10 installation, nor my friend with windows 11 whom I asked to try the script.

I've changed the encoding type to ascii, but there was strange behavior I encountered during testing which could have been fixed with `-Encoding bigendianunicode`.

Can you test this out? I will explore the solutions in that stackoverflow post too, but let me know if you figure anything out before I make changes.

[the remark this answers is not in the document]
Comment (bmFtZQ) [reply install.ps1]: If makes any difference, this error occurred on a Windows 11 on ARM VM running on my M1 Mac, with the ARM64 version of Firefox.

When the encoding was changed to ASCII, the file worked without issues but any other encoding and Firefox wouldn't accept the file.

From that Stack Overflow thread, the answers that used `[System.IO.File]::WriteAllLines()` worked and creates the file without the BOM, but this will overwrite everything in the file.

Oddly enough, the append version of this function: `[System.IO.File]::AppendAllLines()` does not work the same way with this error output to the console:

```
Cannot find an overload for "AppendAllLines" and the argument count: "2".
```

But then I found this GitHub Gist that shows it can be used in PS if its inputs are arrays: https://gist.github.com/refactorsaurusrex/9aa6b72f3519dbc71f7d0497df00eeb1

Applying this function to the script, results in the output file encoded as UTF-8 without the BOM (and Firefox accepts the file):

Replace Line 21 with:
```ps
$path="$PROFILE_ROOTDIR\user.js"
$output=(Write-Output "user_pref(`"$pref`", $bool);" | % { $_ -replace ' +$','' })
[System.IO.File]::AppendAllLines([string]$path, [string[]]$output)
```


Replace Line 144 with:
```ps
$path="$PROFILE_ROOTDIR\user.js"
$output=(Get-Content "$env:temp\$PROJECT_NAME-main\user.js")
[System.IO.File]::AppendAllLines([string]$path, [string[]]$output)
```

**Comment by DanielSGH** [comment]
@bmFtZQ 
> Also with the user.js file, if running the script again after installing it, duplicate values will be added to the user.js file as Firefox will never edit user.js. When Firefox is opened, it sets the preferences found in user.js. Removing them from the user.js file later does not remove the preferences from Firefox as they have already been set previously.

The preferences that firefox sets are found in prefs.js, which means they can easily be removed (which sets them to default) provided firefox is closed. Of course, this would mean we also remove the prefs in user.js, because we don't want to set them again. 

TL;DR change nothing about how it's currently being installed, but on uninstall, remove prefs in user.js and prefs.js.

The option is there, let me know if you prefer that instead.

**Comment by bmFtZQ** [comment]
> The preferences that firefox sets are found in prefs.js, which means they can easily be removed (which sets them to default) provided firefox is closed. Of course, this would mean we also remove the prefs in user.js, because we don't want to set them again.
> 
> TL;DR change nothing about how it's currently being installed, but on uninstall, remove prefs in user.js and prefs.js.
> 
> The option is there, let me know if you prefer that instead.

When uninstalling I think removing the preferences should be optional and the script could prompt the user with something like this: 

```
Reset preferences to default values in 'prefs.js' and 'user.js'? [y/N]
```

Then also create a backup of the existing files before starting (create a copy with `.bak` added to the end of the filename).

When installing, maybe the script can look in the `user.js` file for existing preferences and only set the preferences that are needed.

If someone's existing `user.js` file looked like this:

```js
user_pref("toolkit.legacyUserProfileCustomizations.stylesheets", true);
user_pref("layout.css.color-mix.enabled", true);
user_pref("layout.css.light-dark.enabled", true);
```

then only `user_pref("svg.context-properties.content.enabled", true);` would need to be added to the file.

File: install.sh [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+# UTILITY FUNCTIONS
+set_pref() {
+  echo "setting $1 to $2 in user.js";
+  echo "user_pref(\"$1\", $2);" >> $PROFILE_ROOTDIR/user.js;
+}
+
+delete_pref() {
+  echo "resetting $(echo $1 | cut -d '"' -f 2) to default";
+  sed -i "/$1/d" "$PROFILE_ROOTDIR/user.js"
```
Comment (bmFtZQ) [inline install.sh:34]: This causes an error on macOS:

```
sed: 1: command a expects \ followed by text
```

Adding ` ''` after `-i` seems to fix it, but breaks when run on Linux with the GNU version of `sed`.

Here is a Stack Overflow question that contains more information: https://unix.stackexchange.com/questions/92895/how-can-i-achieve-portability-with-sed-i-in-place-editing

**Review by bmFtZQ — CHANGES_REQUESTED** [review]
The Windows version works great, though the `widget.macos.native-context-menus` preference isn't needed there as it would have no effect on Windows.

With the bash script, there are a few changes (mostly related to adding quotes around path names), but also an annoying bug due to differences in the BSD and GNU versions of `sed`.

File: install.sh [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  echo -e "${RED}Installation failed: ${NC}Failed to get github repo tarball and extract it";
+  exit 0;
+fi
+
+echo "Copying theme folder...";
+cp -r $TMP_DIR/chrome $PROFILE_ROOTDIR;
+cat $TMP_DIR/user.js | 
+  while read line; do
+    if ! grep -qF "$line" $PROFILE_ROOTDIR/user.js; then
+      echo "$line" | tee -a $PROFILE_ROOTDIR/user.js >/dev/null;
```
[the remark this answers is not in the document]
Comment (DanielSGH) [reply install.sh:148]: It was for testing purposes at the time, since `tee` directs output to the file and `stdout`, therefore very useful for debug logs. It was easier to stop output with `>/dev/null` than it was to change the whole command.

It's better to change it to `echo $contents >> $file` now though.

File: install.sh [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    fi;
+  done;
+
+#####################
+# OPTIONAL SETTINGS #
+#####################
+
+echo "Optional settings, refer to https://github.com/bmFtZQ/edge-frfox/tree/main#how-to-install";
+for ((i = 0; i < ${#OPTIONALS[@]}; i += 2)); do
+  if grep -qF "user_pref(\"${OPTIONALS[i]}\", ${OPTIONALS[i+1]});" $PROFILE_ROOTDIR/user.js; then
```
[the remark this answers is not in the document]
Comment (DanielSGH) [reply install.sh:158]: The error doesn't seem to appear for me, but maybe it does appear on macOS if that's what you're testing it on. I'll add a fix

File: install.sh [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+  for ((i = 0; i < ${#FETCHED_PREFS[@]}; i++)); do
+    delete_pref "${FETCHED_PREFS[i]}" $reset_all;
+  done;
+
+  ans="y"
+  echo -e "${CYAN}NOTE: enter 'a' to answer 'yes' to all questions.${NC}";
+  for ((i = 0; i < ${#OPTIONALS[@]}; i+=2)); do
+    setting="user_pref(\"${OPTIONALS[i]}\", ${OPTIONALS[i+1]});";
+
+    if [[ ! ${ans,,} =~ ^(all|a)$ ]]; then
```
[the remark this answers is not in the document]
Comment (DanielSGH) [reply install.sh:119]: `ask_question` returns either true or false for 'yes' or 'no' answers respectively, in order to adapt this code to work with ask_question, I'd have to remove the possibility of answering 'all'. Doing it any other way kind of makes the existence of `ask_question` pointless, since it wouldn't really shorten the code. Perhaps I should rename the function to `ask_binary_question` though, since that might be more descriptive.

[the remark this answers is not in the document]
Comment (bmFtZQ) [reply install.sh]: The name of the function is probably fine. Not using the `ask_question` function would just mean that the `tr` command will need to be called again (if you want to accept both upper and lower case inputs).

Maybe that could be turned into a function if the conversion is needed enough times?

Something like this:

```sh
to_lowercase() {
  echo $1 | tr "[:upper:]" "[:lower:]"
}
```

Then it can be called like this:
```sh
$(to_lowercase $input)
```

[the remark this answers is not in the document]
Comment (DanielSGH) [reply install.sh]: > Not using the ask_question function would just mean that the tr command will need to be called again (if you want to accept both upper and lower case inputs).

That's already been done, check latest changes. I'll change it into a function now too, why not.

**Comment by DanielSGH** [comment]
Happy New Year!

**Comment by bmFtZQ** [comment]
Happy new year!

**Comment by DanielSGH** [comment]
@bmFtZQ PR is ready to merge unless you want anything else changed.

File: install.sh [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+  else
+    echo -e "${YELLOW}WARNING: Failed to retrieve online user.js prefs. Continuing...${NC}"
+  fi;
+
+  for ((i = 0; i < ${#FETCHED_PREFS[@]}; i++)); do
+    delete_pref "${FETCHED_PREFS[i]}" $reset_all;
+  done;
+
+  ans="y"
+  echo -e "${CYAN}NOTE: enter 'a' to answer 'yes' to all questions.${NC}";
```
Comment (bmFtZQ) [inline install.sh:124]: The following section does not seem to work in either Linux or macOS it shows the prompt, but then nothing happens. The next line printed to the terminal is `uninstall complete`.

I think this section can just be removed, deleting the contents of `user.js` can just be handled by the other section (if the user input yes for that question).

File: install.ps1 [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+  if (!$fail) {
+    $FETCHED_PREFS= @()
+    $FETCHED_PREFS += Select-String -Path "$env:temp\user.js" -Pattern "user_pref\(`"([a-zA-Z0-9.-]+)`",\s*(true|false)\);" | ForEach-Object {$_.Line -replace ".*\\", ""}
+    foreach ($pref in $FETCHED_PREFS) {
+      delete_pref "$pref" $reset_all;
+    }
+  } 
+
+  $ans = "y"
+  Write-Host "NOTE: enter 'a' to answer 'yes' to all questions." -ForegroundColor Cyan
```
Comment (bmFtZQ) [inline install.ps1:110]: Same thing as mentioned in the bash shell script.

File: install.ps1 [hunk 16]
```diff
@@ -0,0 +1,186 @@
+# Get-ExecutionPolcy = "Restricted" by default
+
+# VARIABLES, CHANGE AS NEEDED
+$GITHUB_REPO="https://github.com/bmFtZQ/edge-frfox.git"
+$PROJECT_NAME=$($GITHUB_REPO.Split("/")[-1] -replace ".git", "")
+$FIREFOX_DIR="$env:APPDATA\Mozilla\Firefox";
+$PROFILE_ROOTDIR="$($FIREFOX_DIR)\Profiles\$((Select-String -Path "$($FIREFOX_DIR)\profiles.ini" -Pattern "Path=.*\.(dev-edition-default|default-.*)" | Select-Object -Last 1).Line.Substring(14))";
+$OPTIONALS=@{
+  "browser.theme.dark-private-windows" = "false"
```
Comment (bmFtZQ) [inline install.ps1:9]: While we're here, add the new optional preference for Windows: `browser.theme.windows.accent-color-in-tabs.enabled` = `true`.

File: install.sh [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+if [[ $OSTYPE == "darwin"* ]]; then
+  FIREFOX_DIR="$HOME/Library/Application Support/Firefox"
+else
+  FIREFOX_DIR=$HOME/.mozilla/firefox
+fi;
+
+PROFILE_ROOTDIR=$FIREFOX_DIR/$(grep -E "Path=.*\.(dev-edition-default|default-.*)" "$FIREFOX_DIR/profiles.ini" | tail -1 | cut -c 6-);
+OPTIONALS=(
+  "widget.macos.native-context-menus" "false"
+  "browser.theme.dark-private-windows" "false"
```
Comment (bmFtZQ) [inline install.sh:16]: Same as in the powershell script, but with `widget.gtk.rounded-bottom-corners.enabled` = `true` instead.

**Review by bmFtZQ — CHANGES_REQUESTED** [review]
Two final changes, first remove/fix the section that asks you if you want to remove preferences. I think it is unnecessary with the option to remove preferences from `user.js` and `prefs.js`.

The second change is just adding some new additional preference values. (One for Linux, and one for Windows.)
